Week7/app/src/main/java/com/bartoszkrol/catfacts/App.kt

```diff
@@ -0,0 +1,22 @@
+package com.bartoszkrol.catfacts
+
+import android.app.Application
+import com.bartoszkrol.catfacts.networking.RemoteApi
+import com.bartoszkrol.catfacts.networking.buildApiService
+
+class App : Application() {
+
+    companion object {
+        private lateinit var instance: App
+
+        private val apiService by lazy { buildApiService() }
+
```

Comment:
Would love to see the DAO and Repository here too

Reply:
I couldn't create a repository inside the application class. I think it's because database is no't ready at the start (problem with synchronization).

Reply:
Well you can also use a lazy value, and have the context passed in be the `instance`.

It should work 100%, as that's usually the route I take!

Reply:
I have already tried using lazy just like in apiService. I get this error "lateinit property catdao has not been initialized". I am passing context from the instance class, i also tried to use lateinit var and init this classes inside onCreate method.

Reply:
Well if you add it to the App class, and have everything as a lazy-delegate-value, including the DAO, you should be fine.

Reply:
val catdao by lazy {
            CatFactsDatabase.getDatabase(instance).catFactsDao()
        }
        
        val repository: RoomRepository by lazy {
            RoomRepository(catdao)
        }

Reply:
Caused by: kotlin.UninitializedPropertyAccessException: lateinit property instance has not been initialized
        at com.bartoszkrol.catfacts.App$Companion$catdao$2.invoke(App.kt:19)
        at com.bartoszkrol.catfacts.App$Companion$catdao$2.invoke(App.kt:11)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at com.bartoszkrol.catfacts.App$Companion.getCatdao(App.kt)
        at com.bartoszkrol.catfacts.App$Companion$repository$2.invoke(App.kt:23)
        at com.bartoszkrol.catfacts.App$Companion$repository$2.invoke(App.kt:11)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at com.bartoszkrol.catfacts.App$Companion.getRepository(App.kt)
        at com.bartoszkrol.catfacts.viewmodel.CatFactsViewModel.<init>(CatFactsViewModel.kt:15)

Reply:
How do you use the repository in the ViewModel though?

Reply:
Inside the CatFactsViewModel:

private val repository: RoomRepository = App.repository
...
fun getCatFacts() = repository.getAllCatFacts()
...

Reply:
You should either pass it in the constructor, or use a lazy delegate here too. Never use direct assigning, because that's memory hogging, and if you don't reach that screen/part of the app, it's just going to take up memory right away.

There might be another issue in the setup, but from the crash, it's coming from the ViewModel!

Reply:
I removed this field and changed it to repository directly from the application class.

class CatFactsViewModel(application: Application) : AndroidViewModel(application) {

    fun getCatFacts() = App.repository.getAllCatFacts()

    fun removeCatFacts() = viewModelScope.launch {
        App.repository.removeAllCatFacts()
    }

    fun insertCatFacts(catFacts: List<CatFact>) = viewModelScope.launch {
        App.repository.insertCatFacts(catFacts)
    }

}

Reply:
Still got an error

   Caused by: kotlin.UninitializedPropertyAccessException: lateinit property instance has not been initialized
        at com.bartoszkrol.catfacts.App$Companion$catdao$2.invoke(App.kt:19)
        at com.bartoszkrol.catfacts.App$Companion$catdao$2.invoke(App.kt:11)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at com.bartoszkrol.catfacts.App$Companion.getCatdao(App.kt)
        at com.bartoszkrol.catfacts.App$Companion$repository$2.invoke(App.kt:23)
        at com.bartoszkrol.catfacts.App$Companion$repository$2.invoke(App.kt:11)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at com.bartoszkrol.catfacts.App$Companion.getRepository(App.kt)
        at com.bartoszkrol.catfacts.viewmodel.CatFactsViewModel.getCatFacts(CatFactsViewModel.kt:13)
        at com.bartoszkrol.catfacts.ui.main.MainActivity.onCreate(MainActivity.kt:40)
        at android.app.Activity.performCreate(Activity.java:6237)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369)

Reply:
That's also not good, because you won't be able to test this.

I did mention the best way to go about this is to use the constructor to send the repository, and not having to create it here. Try if you can do it that way!

Reply:
How can i pass the constructor to the viewmodel when i am initializing it like this 
     ViewModelProvider(this).get(CatFactsViewModel::class.java)

Reply:
https://medium.com/koderlabs/viewmodel-with-viewmodelprovider-factory-the-creator-of-viewmodel-8fabfec1aa4f

We linked this article in the assignments, you should be able to use the factory to pass in everything.

Let me know if this works!

Reply:
I have created this factory for my viewmodel but the error still remains the same. Right now i am debugging to get more information why this isn't working.

Reply:
I will try to check out the project a bit later and check it out for you too, if you don't find out the issue cause

Reply:
Yey! I found the solution :)

Reply:
Debugging the app really helped me a lot! I found that my properties inside the application class weren't ready before my activity was created.  I checked manifest file and i was missing my application class tag

android:name=".App"

Reply:
Awesome! Glad we worked it out! :]

Reply:
Thanks for your help filip!
I have fixed the code so I can merge PR